client/core/bookie.go

```diff
@@ -91,15 +94,16 @@ func (c *candleCache) init(in []*msgjson.Candle) {
 	}
 }
 
-// addCandle adds the candle using candles.Cache.Add.
-func (c *candleCache) addCandle(msgCandle *msgjson.Candle) *msgjson.Candle {
+// addCandle adds the candle using candles.Cache.Add. It returns most recent candle
+// in cache.
+func (c *candleCache) addCandle(msgCandle *msgjson.Candle) (recent msgjson.Candle, ok bool) {
```

Comment:
Why are we changing the return type? Just return a pointer.

Reply:
I find pointers are harder to think about (specifically when they appear in concurrent code), so just follow a convention of trying to limit the scope of their usage when possible. Could change it back to pointer - it's not really affecting anything here.

Reply:
I'm still trying to wrap my head around the issue, but pointers are passed around to save on memory allocations and speed up the app. And, it can be just as confusing to copy structs. Take this for example, some types from the copied struct are not actually copied but passed by reference themselves: https://go.dev/play/p/TAYp0ruB6pL

Reply:
interfaces also like to sneak past our intuition as under the hood they are basically data pointers with a type label in tow.  Fun fact - maps (if initialized) are literally pointers to unexported types the implementation. https://dave.cheney.net/2017/04/30/if-a-map-isnt-a-reference-variable-what-is-it

Still, there are good points in this discussion that it's arguably more expressive to be returning a value.  However, I don't think we should follow that rigidly at all, especially with objects that are many times larger than a pointer, but there are merits.

Reply:
> And, it can be just as confusing to copy structs.

True, but all else being equal, I'd use whatever primitives language provides to lower the probability of making error.

> However, I don't think we should follow that rigidly at all, especially with objects that are many times larger than a pointer, but there are merits.

I didn't refactor it because of refactor, changing return type there was the first thing that came to mind to solve the problem and I didn't think it is a change that could rub anyone in a wrong way (it's couple of line, and straightforward). Everyone has different preferences for small details, I too don't think we should be spending much time discussing/fighting over them.